pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go

```diff
@@ -0,0 +1,431 @@
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/ugorji/go/codec"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+	"github.com/Azure/ARO-RP/pkg/frontend/adminactions"
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+	utilcert "github.com/Azure/ARO-RP/pkg/util/cert"
+	utilpem "github.com/Azure/ARO-RP/pkg/util/pem"
+	"github.com/Azure/ARO-RP/pkg/util/steps"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+type etcdrenew struct {
+	log           *logrus.Entry
+	k             adminactions.KubeActions
+	secretNames   []string
+	mode          string
+	backupSecrets map[string][]byte
+	lastRevision  int32
+}
+
+var etcdOperatorControllerConditionsExpected = map[string]operatorv1.ConditionStatus{
+	"EtcdCertSignerControllerDegraded": operatorv1.ConditionFalse,
+}
+
+var etcdOperatorConditionsExpected = map[configv1.ClusterStatusConditionType]configv1.ConditionStatus{
+	configv1.OperatorAvailable: configv1.ConditionTrue,
+	configv1.OperatorDegraded:  configv1.ConditionFalse,
+}
+
+func (f *frontend) postAdminOpenShiftClusterEtcdCertificateRenew(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	err := f._postAdminOpenShiftClusterEtcdCertificateRenew(ctx, r, log)
+	if err != nil {
+		log.Infoln("Done")
+	}
+
+	adminReply(log, w, nil, nil, err)
+}
+
+// validate cluster is <4.9 and etcd operator is in expected state
+// Secrets exists, unexpired and close to expiry
+// backup and delete secrets, if backupAndDelete is set True
+func (e *etcdrenew) validateEtcdAndBackupDeleteSecretOnFlagSet(ctx context.Context, backupAndDelete bool) error {
+	s := []steps.Step{
+		steps.Action(e.validateEtcdOperatorControllersState),
+		steps.Action(e.validateEtcdOperatorState),
+		steps.Action(e.validateEtcdCertsExistsAndExpiry),
+	}
+	if backupAndDelete {
+		s = append(s,
+			steps.Action(e.fetchEtcdCurrentRevision),
+			steps.Action(e.backupEtcdSecrets),
+			steps.Action(e.deleteEtcdSecrets),
+		)
+	}
+
+	_, err := steps.Run(ctx, e.log, 10*time.Second, s, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Etcd secrets are deleted or updated, a new revision is will put and applied
+// This function polls if a new revision is applied successfully
+func (e *etcdrenew) isEtcDRootCertRenewed(ctx context.Context) error {
+	s := []steps.Step{
+		steps.Condition(e.isEtcdRevised, 30*time.Minute, true),
+	}
+	_, err := steps.Run(ctx, e.log, 10*time.Second, s, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *etcdrenew) revertChanges(ctx context.Context) error {
+	s := []steps.Step{
+		steps.Action(e.fetchEtcdCurrentRevision),
+		steps.Action(e.recoverEtcdSecrets),
```

Comment:
We have discussed this with @SrinivasAtmakuri. We decided not to dump the secret for following reasons:

- The code is only short lived, once we have all cluster on 4.9, we can remove it. Meaning no need for over engineering
- There is a KCS, which Srini can link, that has step by step process to mitigate issue when certificate is expired or lost
- Even with moderate risk more cluster would work and not fail

Reply:
@SudoBrendan @petrkotas I think the following KCS: https://access.redhat.com/solutions/7023254 can be applied if we fail during recovery and the secrets go missing. The KCS is applied when all etcd secrets of the cluster are expired and certificates need to be manually created. (P.S. This was not tested)

I have taken some time out today and have added changes to copy/delete the certificates as blobs in "aro" containers of the cluster Storage Account.

The code doesn't try reading the storage account during recovery and fails, but will let SREs perform a manual recovery if this goes sideways.

Please review. Thank you.

Reply:
Also, I have made a few functions as Conditional steps, to have more retries as sugggested.

Reply:
wrt retries -- if we retry unconditionally for up to a minute or so, we should be able to do it fine. We shouldn't have it as Conditions, though, since those are meant to be idempotent.

Reply:
FYI, I have removed the code to save the secrets in the cluster storage account after discussions with @hawkowl @petrkotas